src/main/java/uk/gov/pay/adminusers/service/InviteCompleter.java

```diff
@@ -0,0 +1,66 @@
+package uk.gov.pay.adminusers.service;
+
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import uk.gov.pay.adminusers.model.Invite;
+import uk.gov.pay.adminusers.model.Service;
+import uk.gov.pay.adminusers.persistence.dao.InviteDao;
+import uk.gov.pay.adminusers.persistence.dao.ServiceDao;
+import uk.gov.pay.adminusers.persistence.dao.UserDao;
+import uk.gov.pay.adminusers.persistence.entity.ServiceEntity;
+import uk.gov.pay.adminusers.persistence.entity.UserEntity;
+
+import java.util.Optional;
+
+import static uk.gov.pay.adminusers.service.AdminUsersExceptions.conflictingEmail;
+import static uk.gov.pay.adminusers.service.AdminUsersExceptions.inviteLockedException;
+
+public class InviteCompleter {
+    @Inject
+    private final InviteDao inviteDao;
+    private final UserDao userDao;
+    private final ServiceDao serviceDao;
+    private final LinksBuilder linksBuilder;
+
+    @Inject
+    public InviteCompleter(InviteDao inviteDao, UserDao userDao, ServiceDao serviceDao,  LinksBuilder linksBuilder) {
+        this.inviteDao = inviteDao;
+        this.userDao = userDao;
+        this.serviceDao = serviceDao;
+        this.linksBuilder = linksBuilder;
+    }
+
+    /**
+     * Completes an invite.
+     * ie. it creates and persists a user from an invite
+     * and if it is a service invite also creates a default service.
+     * It then disables the invite.
+     */
+    @Transactional
+    public Optional<Invite> complete(String inviteCode) {
+        return inviteDao.findByCode(inviteCode)
+                .map(inviteEntity -> {
+                    if (inviteEntity.isExpired() || inviteEntity.isDisabled()) {
+                        throw inviteLockedException(inviteEntity.getCode());
+                    }
+
+                    if (userDao.findByEmail(inviteEntity.getEmail()).isPresent()) {
+                        throw conflictingEmail(inviteEntity.getEmail());
+                    }
+
+                    if (inviteEntity.isServiceType()) {
+                        ServiceEntity serviceEntity = ServiceEntity.from(Service.from());
+                        serviceDao.persist(serviceEntity);
+                        inviteEntity.setService(serviceEntity);
```

Comment:
My feeling is that these association(s) are not good. As it is InviteEntity has 2 hard associations to UserEntity and ServiceEntity at the moment. 
My feeling is that we should;
 - treat `Invite` as a completely independent entity not having any hard associations with any other adminuser concepts. 
- Its `OK` to have a 2 invite types  (service/user) but its only to determine what this invite is for but should not know whats the final Service or User its creating.
- "service_id" and "sender_id" should be plain Strings (rather than entities) and they should only be used for
  - "service_id" of the existing service for which this new User is invited to. 
  - "sender_id" the userId of the user who invites the new member. 

Its still okay to have a `mapToUserEntity` and/or `mapToServiceEntity` purely to construct a new User or a ServiceEntity out of this invite, but they are merely utilities for copying stuff rather than modifying the invite entity itself. Any other assignments (assigning new service to a user etc.) must be done in a controller logic area.
Happy to discuss it.